### PR TITLE
[ADD] hr_employee_contract_history

### PR DESCRIPTION
**Pull request by saltonmassally**
Calculates first and current contract for employees and assigns job position
on current contract to employee

### CODE REVIEW

File: hr_employee_contract_chronology/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#
+###############################################################################
+{
+    "name": "Employee Contract Chronology",
+    'version': '8.0.1.0.0',
+    "author": "Salton Massally <smassally@idtlabs.sl>",
+    "website": "http://idtlabs.sl",
+    "category": "Human Resources",
+    "summary": "Small usability improvements for employee contracts",
+    "depends": [
```
Comment (dreispt) [inline hr_employee_contract_chronology/__openerp__.py]: add license key

File: hr_employee_contract_chronology/__init__.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+###############################################################################
+#
+#    Copyright (C) 2016 Salton Massally (<smassally@idtlabs.sl>).
```
Comment (dreispt) [inline hr_employee_contract_chronology/__init__.py]: 2016?

File: hr_employee_contract_chronology/models/hr_employee.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    @api.depends(
+        'contract_ids',
+        'contract_ids.date_start',
+        'contract_ids.employee_id',
+        'contract_ids.job_id'
+    )
+    def _compute_first_contract(self):
+        if len(self.contract_ids):
+            contracts_sorted = self.contract_ids.sorted(
+                key=operator.itemgetter('date_start', 'id'))
```
Comment (dreispt) [inline hr_employee_contract_chronology/models/hr_employee.py]: If you prefer, a arguably simpler alternative is: `key=lambda r: (r.date_start, r.id)`

File: hr_employee_contract_chronology/README.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+============================
+Employee Contract Chronology
```
Comment (dreispt) [inline hr_employee_contract_chronology/README.rst]: I would name this `Employee Contract History`

File: hr_employee_contract_chronology/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#    along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+{
+    "name": "Employee Contract Chronology",
+    'version': '8.0.1.0.0',
+    "author": "Salton Massally <smassally@idtlabs.sl>",
+    "website": "http://idtlabs.sl",
+    "category": "Human Resources",
+    "summary": "Small usability improvements for employee contracts",
```
Comment (dreispt) [inline hr_employee_contract_chronology/__openerp__.py]: "Manage employee contract history"

File: hr_employee_contract_chronology/models/hr_contract.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+from openerp import fields, models
+
+
+class HrContract(models.Model):
+    _inherit = 'hr.contract'
+
```
Comment (dreispt) [inline hr_employee_contract_chronology/models/hr_contract.py]: add `_order = "date_start desc, id desc"`

File: hr_employee_contract_chronology/models/hr_employee.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    contract_id = fields.Many2one(
+        'hr.contract',
+        'Current Contract',
+        help='Latest contract of the employee',
+        store=True,
+        readonly=True,
+        index=True,
+        compute='_compute_current_contract'
+    )
```
Comment (ghost) [inline hr_employee_contract_chronology/models/hr_employee.py]: you don't need to recompute the whole field, only add the arguments that change

File: hr_employee_contract_chronology/models/hr_employee.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        'contract_ids',
+        'contract_ids.date_start',
+        'contract_ids.employee_id',
+        'contract_ids.job_id'
+    )
+    def _compute_current_contract(self):
+        if len(self.contract_ids):
+            contracts_sorted = self.contract_ids.sorted(
+                key=lambda r: (r.date_start, r.id))
+            self.contract_id = contracts_sorted[-1]
```
Comment (ghost) [inline hr_employee_contract_chronology/models/hr_employee.py]: I'm curious about these functions. please add tests. it should be systematic. You want to publish a module in OCA, it should test what it does.

File: hr_employee_contract_chronology/models/hr_employee.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+    @api.one
+    @api.depends(
+        'contract_ids',
+        'contract_ids.date_start',
+        'contract_ids.employee_id',
+        'contract_ids.job_id'
+    )
+    def _compute_current_contract(self):
+        if len(self.contract_ids):
```
Comment (ghost) [inline hr_employee_contract_chronology/models/hr_employee.py]: len is redondant

File: hr_employee_contract_chronology/models/hr_employee.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+    @api.one
+    @api.depends(
+        'contract_ids',
+        'contract_ids.date_start',
+        'contract_ids.employee_id',
+        'contract_ids.job_id'
+    )
+    def _compute_first_contract(self):
+        if len(self.contract_ids):
```
Comment (ghost) [inline hr_employee_contract_chronology/models/hr_employee.py]: len is redondant

File: hr_employee_contract_chronology/README.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+==========================
+Employee Contract History
+==========================
+
+Calculates first and current contract for employees and assigns job position
+on current contract to employee
```
Comment (ghost) [inline hr_employee_contract_chronology/README.rst]: Sorry, this is not very clear to me. Please detail a use case of this feature.

File: hr_employee_contract_chronology/models/hr_employee.py (hunk 7 above)
Comment (dreispt) [inline hr_employee_contract_chronology/models/hr_employee.py]: Isn't that defined in the method's `@depends` decorator?

File: hr_employee_contract_history/__openerp__.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+{
+    "name": "Employee Contract History",
+    'version': '8.0.1.0.0',
+    'license': 'AGPL-3',
+    "author": "Salton Massally <smassally@idtlabs.sl>",
```
Comment (feketemihai) [inline hr_employee_contract_history/__openerp__.py:24]: Please add OCA as author of the module...

File: hr_employee_contract_history/models/hr_employee.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        store=True,
+        related='contract_id.job_id'
+    )
+
+    @api.one
+    @api.depends(
+        'contract_ids',
+        'contract_ids.date_start',
+        'contract_ids.employee_id',
+        'contract_ids.job_id'
```
Comment (feketemihai) [inline hr_employee_contract_history/models/hr_employee.py:51]: I don't think the employee_id and job_id should be dependants, if you change the employee the contract will not be in the contract_ids, and even if you change the job_id, it's just adding a recomputation with the same value.

File: hr_employee_contract_history/models/hr_employee.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+            contracts_sorted = self.contract_ids.sorted(
+                key=lambda r: (r.date_start, r.id))
+            self.first_contract_id = contracts_sorted[0]
+
+    @api.one
+    @api.depends(
+        'contract_ids',
+        'contract_ids.date_start',
+        'contract_ids.employee_id',
+        'contract_ids.job_id'
```
Comment (feketemihai) [inline hr_employee_contract_history/models/hr_employee.py:64]: In this case you don't need the dependant on the employee_id, just like in the upper comment.

File: hr_employee_contract_history/tests/test_hr_employee_contract_history.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+                'name': 'Contract 1',
+                'date_start': '2012-01-01',
+                'date_end': '2012-12-31',
+                'wage': 5000
+            }
+        )
+        self.contract_2 = self.contract_model.create(
+            {
+                'employee_id': self.employee.id,
+                'name': 'Contract 1',
```
Comment (feketemihai) [inline hr_employee_contract_history/tests/test_hr_employee_contract_history.py:46]: Rename the contract as Contract 2, in the code you added the job_id required, so probably it's usefull to add the job_id data in the test, probably to add a test if the job_id from the employee is changing when you change the job_id of the last contract...

**Comment by dreispt** [comment]
@tarzan0820 I think there are comments for you to address.

File: hr_employee_contract_history/models/hr_employee.py (hunk 13 above)
Comment (saltonmassally) [inline hr_employee_contract_history/models/hr_employee.py:51]: I agree with you on this @feketemihai will make the changes

File: hr_employee_contract_history/tests/test_hr_employee_contract_history.py (hunk 15 above)
Comment (saltonmassally) [inline hr_employee_contract_history/tests/test_hr_employee_contract_history.py:46]: @feketemihai noted, will make the changes

File: hr_employee_contract_history/README.rst [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+None
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/hr/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/hr/issues/new?body=module:%20hr_employee_contract_chronology%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (feketemihai) [inline hr_employee_contract_history/README.rst:41]: The module named was changed to hr_employee_contract_history, please fix it in the issue link

File: hr_employee_contract_history/README.rst [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+Configuration
+=============
+
+This module does not require any special configuration apart from enabling it as
+shown above.
+
+Usage
+=====
+None
+
```
Comment (feketemihai) [inline hr_employee_contract_history/README.rst:33]: Please add Try me on Runbot section...

**Comment by feketemihai** [comment]
@tarzan0820 Do you think it's more appropiate to include this features in the hr_employee_seniority, it will make the module more complex and as roadmap to have a direct page in employee view of contracts history...

**Pull request closed without merging**